### PR TITLE
Private key auth

### PR DESCRIPTION
**Pull request by pswedrak**

### CODE REVIEW

File: idact/core/auth.py [hunk 1]
```diff
@@ -15,9 +15,12 @@ class AuthMethod(Enum):
         :attr:`.PUBLIC_KEY`: Generate a private and public key pair, and
                              install the public key.
 
+         :attr:`.PRIVATE_KEY`: Authenticate with previously generated private key
+
     """
     ASK = 0
     PUBLIC_KEY = 1
+    PRIVATE_KEY = 2
```
Comment (kapiekarz) [inline idact/core/auth.py:23]: I suggest changing the names of both of them to `GENERATE_KEY` and `ASSIGN_KEY` or something better.

File: idact/detail/add_cluster_app/main.py [hunk 2]
```diff
@@ -29,13 +29,15 @@
...
+              help="Authentication method. "
+              "Avilable values: GENERATE_NEW_KEY, PRIVATE_KEY, ASK_EVERYTIME. "
+              "Default: GENERATE_NEW_KEY")
 @click.option('--key',
-              default=KeyType.RSA,
-              type=KeyType,
-              help="Specified key type to be generated. "
+              default="~/.ssh",
+              type=str,
+              help="Previously generated private key "
```
Comment (kapiekarz) [inline idact/detail/add_cluster_app/main.py]: Change the doc for something like `Path to the previously generated private key`

File: idact/detail/add_cluster_app/main.py [hunk 3]
```diff
@@ -29,13 +29,15 @@
               type=int,
               help="The ssh port. Default: 22")
 @click.option('--auth',
-              default=AuthMethod.PUBLIC_KEY,
-              type=AuthMethod,
-              help="Authentication method. Default: AuthMethod.PUBLIC_KEY")
+              default='GENERATE_NEW_KEY',
+              type=str,
+              help="Authentication method. "
+              "Avilable values: GENERATE_NEW_KEY, PRIVATE_KEY, ASK_EVERYTIME. "
```
Comment (grzanka) [inline idact/detail/add_cluster_app/main.py]: Typo in `avilable`

File: idact/detail/auth/authenticate.py [hunk 4]
```diff
@@ -111,7 +111,7 @@ def authenticate(host: str,
                                     after authentication.
 
     """
-    if config.auth not in [AuthMethod.ASK, AuthMethod.PUBLIC_KEY]:
+    if config.auth not in [AuthMethod.ASK, AuthMethod.PUBLIC_KEY, AuthMethod.PRIVATE_KEY]:
```
Comment (grzanka) [inline idact/detail/auth/authenticate.py]: Check how to work with Enums:
https://docs.python.org/3/library/enum.html

Comment (grzanka) [reply idact/detail/auth/authenticate.py]: Consider using `__members__`

File: idact/detail/add_cluster_app/main.py [hunk 5]
```diff
@@ -29,11 +29,21 @@
               type=int,
               help="The ssh port. Default: 22")
 @click.option('--auth',
-              default='PUBLIC_KEY',
+              default='GENERATE_NEW_KEY',
```
Comment (jchyzy) [inline idact/detail/add_cluster_app/main.py]: It should be `GENERATE_KEY`

File: idact/detail/add_cluster_app/main.py [hunk 6]
```diff
@@ -29,11 +29,21 @@
...
-              "Avilable values: PUBLIC_KEY, ASK_EVERYTIME. "
-              "Default: PUBLIC_KEY")
+              "Available values: GENERATE_KEY, PRIVATE_KEY, ASK_EVERYTIME. "
+              "Default: GENERATE_KEY")
+@click.option('--key',
+              default="~/.ssh",
+              type=str,
+              help="Path to the previously generated private key "
+                   "Default location: ~/.ssh")
+@click.option('--key_type',
```
Comment (jchyzy) [inline idact/detail/add_cluster_app/main.py:42]: For now, when only RSA is supported, this option makes no sense. I know that that is for future improvement, but it wouldn't be hard to provide user with this option when other key types would be supported. Or do you think that it should stay anyway?

File: idact/detail/add_cluster_app/main.py [hunk 7]
```diff
@@ -74,22 +86,37 @@ def main(cluster_name: str,
...
-        raise ValueError("Auth must be one of: PUBLIC_KEY, ASK_EVERYTIME")
-
-    cluster = add_cluster(name=cluster_name,
-                          user=user,
-                          host=host,
-                          port=port,
-                          auth=auth_method,
-                          key=key_type,
-                          install_key=install_key)
+        cluster = add_cluster(name=cluster_name,
```
Comment (jchyzy) [inline idact/detail/add_cluster_app/main.py:115]: I would suggest only one `add_cluster` call, without if-else. I think that two variables - `key` and `key_type` are unecessary and only one would be ok.

File: idact/detail/auth/authenticate.py [hunk 8]
```diff
@@ -111,7 +111,7 @@ def authenticate(host: str,
                                     after authentication.
 
     """
-    if config.auth not in [AuthMethod.ASK, AuthMethod.PUBLIC_KEY]:
+    if config.auth not in AuthMethod.__members__.values():
```
Comment (jchyzy) [inline idact/detail/auth/authenticate.py:114]: I think that only one option when this condition can be met is when config.auth is None. In this case error saying that this is not implemented is quite not informative.

File: idact/detail/add_cluster_app/main.py [hunk 9]
```diff
@@ -74,22 +86,39 @@ def main(cluster_name: str,
...
 
-    if auth == 'PUBLIC_KEY':
-        auth_method = AuthMethod.PUBLIC_KEY
-        key_type = KeyType.RSA
+    if auth == 'GENERATE_KEY':
+        auth_method = AuthMethod.GENERATE_KEY
+        if key_type == 'RSA':
+            key_type = KeyType.RSA
+        else:
+            ValueError("Unknown key type")
```
Comment (kapiekarz) [inline idact/detail/add_cluster_app/main.py]: I think here should be rise

File: idact/core/auth.py [hunk 10]
```diff
@@ -12,12 +12,15 @@ class AuthMethod(Enum):
 
         :attr:`.ASK`: Ask for password every time it's needed.
 
-        :attr:`.PUBLIC_KEY`: Generate a private and public key pair, and
+        :attr:`.GENERATE_KEY`: Generate a private and public key pair, and
                              install the public key.
 
+         :attr:`.PRIVATE_KEY`: Authenticate with previously generated key
```
Comment (kapiekarz) [inline idact/core/auth.py]: If you would add the new commit delete one space at the start please.

File: idact/detail/auth/authenticate.py (hunk 8 above)
Comment (pswedrak) [reply idact/detail/auth/authenticate.py:114]: https://en.wikipedia.org/wiki/Assertion_(software_development)

**Review by kapiekarz — APPROVED** [review]

**Pull request merged**